### ISSUE TITLE
Installed libssl.so and libcrypto.so symlinks broken?

### ISSUE THREAD

**Issue by ejparz** [openssl/openssl#16012]
Hi OpenSSL Team,

We were previously using **openssl-3.0.0-alpha9**, and successfully compiling/building the project using the following commands

```
Download and unpack openssl-3.0.0-alpha9 release from github
cd openssl
export ANDROID_NDK_ROOT=$(abspath ../standalone-toolchain) 
export TOP = $(shell pwd)
./Configure android-arm -march=armv7-a --prefix=$(TOP)/install
make
make install
```

You can see the install/lib directory is correctly populated with libssl.so and libcrypto.so:
![image](https://user-images.githubusercontent.com/8854686/124652742-75409b00-de6a-11eb-903e-52e180900a8c.png)

We recently updated to **openssl-3.0.0-beta1**, and used the following commands to compile
```
Download and unpack openssl-3.0.0-beta1 release from github
cd openssl
export ANDROID_NDK_ROOT=$(abspath ../standalone-toolchain) 
export TOP = $(shell pwd)
./Configure android-arm -march=armv7-a --prefix=$(TOP)/install enable-fips
make
make install
```

OpenSSL **openssl-3.0.0-beta1** builds fine using the above commands. 

However, libssl.so and libcrypto.so are not properly installed. (See screenshot below- notice the symlinks are broken)
![image](https://user-images.githubusercontent.com/8854686/124651869-758c6680-de69-11eb-9a8b-75c54844018c.png)

Please see output from 

``` 
perl configdata.pm --dump
```
attached

[alpha_9_config_dump.txt](https://github.com/openssl/openssl/files/6772193/alpha_9_config_dump.txt)
[beta_1_config_dump.txt](https://github.com/openssl/openssl/files/6772194/beta_1_config_dump.txt)

Can you please advise why libcrypto.so and libssl.so are not properly created in **openssl-3.0.0-beta1** ?

Thank you,

Eric

**Comment by t8m** [comment]
If you try `grep 'libcrypto\.so:' Makefile` on a configured build tree what does it print?

**Comment by ejparz** [comment]
> grep 'libcrypto\.so:' Makefile

```
eparziale@lengep1:~/repo_pie/vendor/persistent/openssl$ grep 'libcrypto\.so:' Makefile
libcrypto.so: crypto/aes/libcrypto-shlib-aes-armv4.o \
```

**Comment by rsbeckerca** [comment]
I'm sad to report that NonStop (both platforms) has a similar issue, but we do not even get the `libssl.so.3` file (because we can't actually from the linker). I will do what I can to help with the diagnosis.

**Comment by levitte** [comment]
Try #16023

**Comment by ejparz** [comment]
@levitte 
After patching in the changes from https://github.com/openssl/openssl/pull/16023, and running the following,
```
export ANDROID_NDK_ROOT=$(abspath ../standalone-toolchain) 
export TOP = $(shell pwd)
./Configure android-arm -march=armv7-a --prefix=$(TOP)/install enable-fips
make
make install
```

we get the following compilation error
```
${LDCMD:-arm-linux-androideabi-clang} -fPIE -pthread  --sysroot=/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/sysroot -Wa,--noexecstack -Qunused-arguments -Wall -O3 -march=armv7-a -pie   \
	-o test/tls13encryptiontest \
	test/tls13encryptiontest-bin-tls13encryptiontest.o \
	libssl.a test/libtestutil.a libcrypto.a -ldl -pthread 
arm-linux-androideabi-clang  -I. -Iinclude -Iapps/include  -DOPENSSL_NO_KTLS -fPIE -pthread  --sysroot=/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/sysroot -Wa,--noexecstack -Qunused-arguments -Wall -O3 -march=armv7-a -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF crypto/tls13secretstest-bin-packet.d.tmp -MT crypto/tls13secretstest-bin-packet.o -c -o crypto/tls13secretstest-bin-packet.o crypto/packet.c
arm-linux-androideabi-clang  -I. -Iinclude -Iapps/include  -DOPENSSL_NO_KTLS -fPIE -pthread  --sysroot=/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/sysroot -Wa,--noexecstack -Qunused-arguments -Wall -O3 -march=armv7-a -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF ssl/tls13secretstest-bin-tls13_enc.d.tmp -MT ssl/tls13secretstest-bin-tls13_enc.o -c -o ssl/tls13secretstest-bin-tls13_enc.o ssl/tls13_enc.c
arm-linux-androideabi-clang  -I. -Iinclude -Iapps/include  -DOPENSSL_NO_KTLS -fPIE -pthread  --sysroot=/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/sysroot -Wa,--noexecstack -Qunused-arguments -Wall -O3 -march=armv7-a -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF test/tls13secretstest-bin-tls13secretstest.d.tmp -MT test/tls13secretstest-bin-tls13secretstest.o -c -o test/tls13secretstest-bin-tls13secretstest.o test/tls13secretstest.c
rm -f test/tls13secretstest
${LDCMD:-arm-linux-androideabi-clang} -fPIE -pthread  --sysroot=/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/sysroot -Wa,--noexecstack -Qunused-arguments -Wall -O3 -march=armv7-a -L. -pie   \
	-o test/tls13secretstest \
	crypto/tls13secretstest-bin-packet.o \
	ssl/tls13secretstest-bin-tls13_enc.o \
	test/tls13secretstest-bin-tls13secretstest.o \
	-lssl test/libtestutil.a -lcrypto -ldl -pthread 
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-ssl_lib.o): multiple definition of 'ssl_evp_cipher_free'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-ssl_lib.o): multiple definition of 'ssl_evp_md_free'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-ssl_lib.o): multiple definition of 'ssl_handshake_hash'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-ssl_lib.o): multiple definition of 'ssl_log_secret'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-rec_layer_s3.o): multiple definition of 'RECORD_LAYER_reset_read_sequence'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-rec_layer_s3.o): multiple definition of 'RECORD_LAYER_reset_write_sequence'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-statem.o): multiple definition of 'ossl_statem_export_allowed'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-statem.o): multiple definition of 'ossl_statem_export_early_allowed'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-statem.o): multiple definition of 'ossl_statem_fatal'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-statem.o): multiple definition of 'ossl_statem_send_fatal'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-s3_enc.o): multiple definition of 'ssl3_digest_cached_records'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-ssl_ciph.o): multiple definition of 'ssl_cipher_get_evp'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-ssl_ciph.o): multiple definition of 'ssl_cipher_get_evp_cipher'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-ssl_ciph.o): multiple definition of 'ssl_handshake_md'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-ssl_ciph.o): multiple definition of 'ssl_md'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: ./libssl.a(libssl-lib-t1_enc.o): multiple definition of 'tls1_alert_code'
/home/eparziale/repo_pie/vendor/persistent/standalone-toolchain/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: test/tls13secretstest-bin-tls13secretstest.o: previous definition here
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Makefile:28481: test/tls13secretstest] Error 1
make[2]: Leaving directory '/home/eparziale/repo_pie/vendor/persistent/openssl'
make[1]: *** [Makefile:3383: build_sw] Error 2
make[1]: Leaving directory '/home/eparziale/repo_pie/vendor/persistent/openssl'
make: *** [Makefile.ps:14: e2] Error 2
```

Without the patch, the build completes without compilation errors, but the .so files are not symlinked properly

**Comment by rsbeckerca** [comment]
> 
> 
> Try #16023

I can't even build with it:
```
**** FATAL ERROR **** [1596] - linker halting:
   Illegal duplicate definition of the procedure ssl_handshake_hash in
   test/tls13secretstest-bin-tls13secretstest.o and
   ./libssl.a(libssl-lib-ssl_lib.o).
```

**Comment by rsbeckerca** [comment]
I have changed my approach. Started with current `master`, then merged in `fix-16012` `--strategy=recursive`. Making more progress. Will advise.

**Comment by levitte** [comment]
The build info on tls13secretstest was a bit wrong.  It should link with static libraries

**Comment by rsbeckerca** [comment]
> 
> 
> The build info on tls13secretstest was a bit wrong. It should link with static libraries

Thanks. Doing a clean build after this.

**Comment by rsbeckerca** [comment]
> 
> 
> The build info on tls13secretstest was a bit wrong. It should link with static libraries

Looks like that is still broken at commit 90de809 when merged into master.

**Comment by levitte** [comment]
I've dropped 90de809, it was not right.

It surprises me that you guys get into an issue, though.  What happens if you start from a clean slate?

**Comment by rsbeckerca** [comment]
> 
> 
> I've dropped [90de809](https://github.com/openssl/openssl/commit/90de809da2f6f5840e5d017ae6fcabdbb93e7d12), it was not right.
> 
> It surprises me that you guys get into an issue, though. What happens if you start from a clean slate?

I'm only using clean slate builds these days. Will revert this one and try again.

**Comment by levitte** [comment]
I figured out why that happens.  New commit that fixes the problem has been pushed

**Comment by rsbeckerca** [comment]
> 
> 
> I figured out why that happens. New commit that fixes the problem has been pushed

Got it and redoing from a clean slate :)

**Comment by ejparz** [comment]
Perfect, I confirmed the latest 2-file change in https://github.com/openssl/openssl/pull/16023 works when patched on top of **openssl-3.0.0-beta1**

Thanks!

**Comment by rsbeckerca** [comment]
Way better. Thanks @levitte
